search/tasks.py

```diff
@@ -35,3 +74,127 @@ def update_document_with_partial(doc_id, partial_data):
 def increment_document_integer_field(doc_id, field_name, incr_amount):
     """Task that makes a request to increment some integer field in an ES document"""
     api.increment_document_integer_field(doc_id, field_name, incr_amount)
+
+
+@app.task(autoretry_for=(RetryException, ), retry_backoff=True, rate_limit='600/m')
+def index_post(api_username, post_id):
+    """
+    Index a post and its comments
+
+    Args:
+        api_username (str): The API username
+        post_id (str): The post string
+    """
+    with wrap_retry_exception(PrawcoreException, PRAWException):
+        api.index_post(api_username, post_id)
+
+
+@app.task(bind=True, autoretry_for=(RetryException, ), retry_backoff=True, rate_limit='600/m')
+def index_channel(self, api_username, channel_name):
+    """
+    Index the channel
+
+    Args:
+        api_username (str): API username
+        channel_name (str): The name of the channel
+    """
+    with wrap_retry_exception(PrawcoreException, PRAWException):
+        from channels.api import Api
+
+        client = Api(User.objects.get(username=api_username))
+        channel = client.get_channel(channel_name)
+        post_ids = [post.id for post in channel.new()]
+
+    raise self.replace(
+        group(
+            index_post.si(api_username=api_username, post_id=post_id) for post_id in post_ids
+        )
+    )
+
+
+@app.task(bind=True)
+def start_recreate_index(self, api_username):
+    """
+    Wipe and recreate index and mapping, and index all items.
+
+    Args:
+        api_username (str): The reddit username to use when crawling data
+    """
+    from channels.api import Api
+
+    user = User.objects.get(username=api_username)
+    conn = get_conn(verify=False)
+
+    # Create new backing index for reindex
+    new_backing_index = make_backing_index_name()
+
+    # Clear away temp alias so we can reuse it, and create mappings
+    clear_and_create_index(index_name=new_backing_index)
+    temp_alias = get_reindexing_alias()
+    if conn.indices.exists_alias(name=temp_alias):
+        # Deletes both alias and backing indexes
+        conn.indices.delete_alias(index="_all", name=temp_alias)
+
+    # Point temp_alias toward new backing index
+    conn.indices.put_alias(index=new_backing_index, name=temp_alias)
+
+    # Do the indexing on the temp index
+    log.info("starting to index all channels, posts and comments...")
+
+    client = Api(user)
+    channel_names = [channel.display_name for channel in client.list_channels()]
+    index_channels = group(
+        index_channel.si(api_username=api_username, channel_name=channel_name)
+        for channel_name in channel_names
+    )
+
+    # Use self.replace so that code waiting on this task will also wait on the indexing and finish tasks
+    raise self.replace(
+        chain(
+            index_channels,
+            finish_recreate_index.si(new_backing_index),
+        )
+    )
+
+
+@app.task
+def finish_recreate_index(backing_index):
```

Comment:
Our other tasks are basically pass-throughs that turn logic in `indexing_api.py` into asynchronous tasks. I think it might be useful to do the same here. At least in development I imagine we could get some use from having this functionality available outside of celery tasks. There's also the issue of having essential index manipulation logic in 2 different places. If there's a compelling reason to leave this logic inside of tasks, or it's just too annoying, let me know